### PR TITLE
Support DNS over TCP

### PR DESCRIPTION
**Pull request by milan-zededa**
In pillar we use customized dialers for DNS resolvers to enforce the
source IP address to use, thus effectively choosing the output network
interface for a given DNS request.
However, currently we expect that DNS servers always respond on the UDP
port, but they may in fact fallback to use TCP (same port number).
This happens for example when the response message size is too large.
In that case we get "mismatched local address type" error.

I observed this with Azure datastore and a private DNS server, so it
make sense to support DNS over TCP at least for Downloader, but for
completeness I also added this for the controller name resolution.

Marking as draft until I find a way of reproducing and testing this locally.
Right now I can observe the "mismatched..." error only in customer logs.

Signed-off-by: Milan Lenco <milan@zededa.com>

### CODE REVIEW

**Comment by giggsoff** [comment]
It seems we should pull changes of zedUpload into pillar with `make proto-vendor`.

**Comment by milan-zededa** [comment]
> It seems we should pull changes of zedUpload into pillar with make proto-vendor.

Correct, I forgot that it is a separate Go package.

File: libs/zedUpload/datastore.go [hunk 1]
```diff
@@ -79,9 +79,16 @@ func httpClientSrcIP(localAddr net.IP, proxy *url.URL) *http.Client {
...
-		// XXX can DNS fallback to TCP? Would get a mismatched address if we do
-		d := net.Dialer{LocalAddr: &localUDPAddr}
-		return d.Dial(network, address)
+		switch network {
+		case "udp":
+			d := net.Dialer{LocalAddr: &localUDPAddr}
+			return d.Dial(network, address)
+		case "tcp":
+			d := net.Dialer{LocalAddr: &localTCPAddr}
+			return d.Dial(network, address)
```
Comment (giggsoff) [inline libs/zedUpload/datastore.go:88]: Should we use `"udp", "udp4", "udp6"` and `"tcp", "tcp4", "tcp6"`?

Comment (milan-zededa) [reply libs/zedUpload/datastore.go:88]: I don't think that it is reachable.
See https://go.dev/src/net/dnsclient_unix.go, function `Resolver.exchange()`, lines 167-172:
```
var networks []string
if useTCP {
	networks = []string{"tcp"}
} else {
	networks = []string{"udp", "tcp"}
}
```

Comment (milan-zededa) [reply libs/zedUpload/datastore.go:88]: But we can add it just in case...

**Comment by eriknordmark** [comment]
@milan-zededa I was assuming there would be some test DNS server which returns a ton of A or AAAA records, but I've yet to find one. What I did find so far is a test setup which returns lots of NS records, but I suspect that will not have any impact on the A/AAAA lookup done here.

FWIW the NS lookup fallback to TCP can be seen with
dig -t ns toomanyns-eth.test.globnix.net @nlns.globnix.net

**Comment by milan-zededa** [comment]
@eriknordmark It turned out that the underlying customer issue was different (https://github.com/lf-edge/eve/pull/2800). Device was missing one of two DNS servers. The other one is not able to resolve controller domain (only private destinations) and God knows what truncated response it returns, which causes Golang resolver to try TCP.
Since we haven't yet actually seen any realistic scenario in which A/AAAA lookup needs TCP to succeed, I would suggest to not introduce this change. There is already a "XXX" comment, so we may just leave it at that until we find a case where support for DNS over TCP is really needed.

**Comment by milan-zededa** [comment]
I'm reopening this PR - the problem with the DNS server truncating UDP response and expecting the client to switch to TCP has been observed again in the customer deployment. We tried to build the latest EVE with this PR included and it indeed fixed their problem, so we will need to merge it and prepare an LTS point release with the patch.

File: libs/zedUpload/datastore.go (hunk 1 above)
Comment (milan-zededa) [reply libs/zedUpload/datastore.go]: Done

**Review by eriknordmark — APPROVED** [review]
LGTM
Do we have some test setup where e.g., there is a ton of A or AAAA records configured for some FQDN which we can use to regression test this?

**Comment by milan-zededa** [comment]
> Do we have some test setup where e.g., there is a ton of A or AAAA records configured for some FQDN which we can use to regression test this?

I will prepare something using Eden-SDN, it should be possible.

**Pull request merged**